### PR TITLE
start new session instead of setpgrp for detaching kernel

### PR DESCRIPTION
**Pull request by minrk**
setpgrp seems to cause problems with children spawning subprocesses. setsid doesn't cause the same problems, but has its own issues.

This is WIP because I need to at least add some tests to exercise various subprocesses-of-subprocesses cases.

cc @takluyver who knows more about terminals and process groups than I do.

### CODE REVIEW

**Comment by takluyver** [comment]
The problem doesn't happen with any subprocess - I think it's something that bash does, and more specifically something related to the `-i` flag: if I take that out, then `Popen()` returns.

My guess is that `bash -i` does something with the controlling TTY. setsid() detaches the process from the controlling tty, which probably prevents bash doing something.

I'm still puzzled as to how bash can cause `Popen()` to hang, though. I'd like to understand this properly before we merge this, even if it is the right fix.

**Comment by minrk** [comment]
I'm *super* puzzled, but I've spent the entire day investigating this and don't have a lot more information than when I started. I'm running out of steam to pursue this one.

It's not strictly `-i`, since I've seen the problem with bash without `-i`, though trivial commands like `bash -c exit` or `sleep 5` work fine.

I really don't understand how the pgrp could affect whatever it is that's failing if **all** of stdout, stderr, and stdin are setup as pipes.

**Comment by takluyver** [comment]
The 'controlling tty' of a process is a separate notion from std(in/out/err) - even if those are all connected to pipes, a process can still get its controlling tty by opening `/dev/tty`. Part of what `setsid()` does is to detach the process from its controlling tty, so opening `/dev/tty` no longer works.

**Comment by minrk** [comment]
So it seems like starting a new session is what we want, then, not setpgrp. I'll update this to use `start_new_session` on py3 and fallback on setsid for Legacy Python.

**Comment by willingc** [comment]
@minrk Nice tests :+1: The logic for all looks good.

**Comment by minrk** [comment]
this now uses `start_new_session` when available, and `setsid` on old and busted Python. I added a test that verifies that the kernel can start `bash -i` subprocesses and the whole process group gets interrupted (the same test passes without the session patch as long as the subprocess is simpler, such as `sleep 30`).

**Comment by jdemeyer** [comment]
The comment is wrong:
```
We don't use setpgrp, which can cause problems for kernels starting subprocesses
```

The problem you're seeing is very specific to `bash -i`, it has nothing to do with running subprocesses in general.

**Comment by minrk** [comment]
@jdemeyer It has also been seen by users trying to start supposedly headless java processes, and `bash -c` *without* `-i`, so it's not confined to `bash -i`.

File: jupyter_client/tests/signalkernel.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        
+
+    def do_execute(self, code, silent, store_history=True, user_expressions=None,
+                   allow_stdin=False):
+        code = code.strip()
+        reply = {'status': 'ok'}
+        if code == 'start':
+            child = Popen(['bash', '-i', '-c', 'sleep 30'], stderr=PIPE)
+            self.children.append(child)
+            reply['pid'] = self.children[-1].pid
```
Comment (takluyver) [inline jupyter_client/tests/signalkernel.py]: We're sticking arbitrary extra fields in the reply? I'm a bit surprised none of our machinery complains about that. Do we want to assume that will always be the case? We could stick them in metadata or `user_expressions`.

File: jupyter_client/tests/signalkernel.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                time.sleep(10)
+            except KeyboardInterrupt:
+                reply['interrupted'] = True
+            else:
+                reply['interrupted'] = False
+        else:
+            reply['status'] = 'error'
+            reply['ename'] = 'Error'
+            reply['evalue'] = code
+            reply['tmraceback'] = ['no such command: %s' % code]
```
Comment (takluyver) [inline jupyter_client/tests/signalkernel.py]: typo in traceback

File: jupyter_client/tests/test_kernelmanager.py [hunk 3]
```diff
@@ -51,3 +82,41 @@ def test_get_connect_info(self):
...
         ])
         self.assertEqual(keys, expected)
+    
+    @dec.skip_win32
+    def test_signal_kernel_subprocesses(self):
+        self._install_test_kernel()
+        km, kc = start_new_kernel(kernel_name='signaltest')
+        def execute(cmd):
+            kc.execute(cmd)
+            reply = kc.get_shell_msg(TIMEOUT)
```
Comment (takluyver) [inline jupyter_client/tests/test_kernelmanager.py]: 30 seconds is going to feel like waiting forever if we ever break what you just fixed and the test kernel is suspended. But I guess any shorter and it risks sporadic failures on Travis.

**Comment by takluyver** [comment]
:+1: overall, just a couple of comments.

**Comment by jdemeyer** [comment]
> @jdemeyer It has also been seen by users trying to start supposedly headless java processes, and bash -c without -i, so it's not confined to bash -i.

Well, the issue that you reported at #68 is certainly confined to `bash -i`: I showed the actual code in the `bash` sources which is responsible for the breakage.

I guess that those other problems you mention are unrelated and perhaps those are not fixed by this PR. Do you have concrete examples?

So I would still suggest to mention in the comments that this PR is meant to fix just `bash -i` and nothing else.

**Comment by willingc** [comment]
@jdemeyer While I understand that you are trying to be precise, the tone of your message is not constructive. "setpgrp" is problematic because its behavior is unpredictable based on OS implementation. The "setpgrp" command may be removed in future standards [reference](http://pubs.opengroup.org/onlinepubs/9699919799/).

**Comment by jdemeyer** [comment]
> the tone of your message is not constructive

Why do you think so? I think you should justify why you said that.

I downloaded the `bash` source code and pointed out which part of `bash` is responsible for the issue mentioned at #68. It shows that the issue is really caused by very specific behaviour of `bash -i`. And then pointing this out is not constructive?

**Comment by jdemeyer** [comment]
> "setpgrp" is problematic because its behavior is unpredictable based on OS implementation.

1) Why do you think that?

2) Why does `setsid()` not have this problem?

> The "setpgrp" command may be removed in future standards reference.

Because it will be replaced by [setpgid()](http://pubs.opengroup.org/onlinepubs/9699919799/functions/setpgid.html), not because process groups are fundamentally problematic.

**Comment by jdemeyer** [comment]
To be clear: I have nothing against replacing `setpgrp()` by `setsid()` (although I'm not particularly in favour of it either).

I am only saying that the very vague comment
```
We don't use setpgrp, which can cause problems for kernels starting subprocesses
```
should be replaced by something like
```
We don't use setpgrp, which can cause problems for running 'bash -i'
```

**Comment by takluyver** [comment]
We observed something that might be similar in jupyter/jupyterhub#382, where it looked like starting another process was causing problems. We didn't investigate that to find out the real cause, though.

@willingc the docs you linked to say:

> applications are encouraged to use setpgid() or setsid() as appropriate.

Any thoughts on which is more likely to be appropriate for us? My gut feeling is that we may as well do setsid - as this PR does - but I don't know this stuff very well.

Shall we make the comment say something like '...starting certain applications in a subprocess'? It's definitely not a problem with starting any subprocess, and it's likely that bash isn't the only application ever written that fires signals around. :-)

**Comment by jdemeyer** [comment]
> applications are encouraged to use setpgid() or setsid() as appropriate.

Right, I missed this part. It's a bit confusing because there is nothing really wrong with process group, only with the `setpgrp()` system call (which should be replaced by `setpgid`).

**Comment by jdemeyer** [comment]
> My gut feeling is that we may as well do setsid - as this PR does - but I don't know this stuff very well.

> Any thoughts on which is more likely to be appropriate for us? My gut feeling is that we may as well do setsid - as this PR does - but I don't know this stuff very well.

Hard to tell, I don't have a strong opinion either way. I cannot really find a major semantic difference between process groups and sessions. Sessions are typically associated to a single interactive login and one might as well consider a Jupyter notebook an "interactive login".

One thing to test is whether this affects shutting down the Jupyter server.

**Comment by takluyver** [comment]
The functional differences I'm aware of:

- `setsid()` detaches the process (and any children) from the controlling terminal. They can still write to the terminal as long as they have an fd for it. This is probably beneficial for us, because it provides a layer of protection against applications trying to grab control of the terminal.
- A process that has called `setsid()` doesn't get sent SIGHUP when its parent process exits (IIUC - this is part of starting a daemon process). When the server shuts down normally, we go through an explicit shutdown process for all the kernels it has started, so this shouldn't make a difference. However, if the notebook server were to crash, presumably its kernels would hang around. That could be good or bad, depending on your point of view.

Is there anything else?

**Comment by jdemeyer** [comment]
> and it's likely that bash isn't the only application ever written that fires signals around. :-)

Sure, but sending signals depending on whether or not the application was started as a separate session is already much more specific.

> Shall we make the comment say something like '...starting certain applications in a subprocess'?

How about "...certain interactive applications (in particular bash -i) in a subprocess"?

**Comment by takluyver** [comment]
That sounds fine to me. :-)

**Comment by willingc** [comment]
@takluyver @jdemeyer I've been traveling all day. 
A few resources:
[Subprocess article (last section on process groups and sessions) by Doug Hellman](https://pymotw.com/2/subprocess/)

I think ``setsid`` fits here. I'm unsure if this should follow the fork/setsid/umask/fork pattern used in a daemon or if setsid would be sufficient. The "Advanced Programming in Unix 3rd ed" by W. Richard Stevens has  excellent chapters on process groups and sessions.

I think we are in agreement that setpgrp is obsolete.
screen shot 2016-01-17 at 6 49 02 pm
screen shot 2016-01-17 at 6 54 27 pm

File: jupyter_client/tests/signalkernel.py (hunk 1 above)
Comment (minrk) [inline jupyter_client/tests/signalkernel.py]: That's me being lazy. With the kernelbase API there's no way to add metadata, so I used user_expressions.

**Comment by minrk** [comment]
Added a note to the effect of `bash -i` and other subprocesses being the reason we don't use setpgrp.

**Comment by willingc** [comment]
@minrk +1

**Comment by takluyver** [comment]
Thanks everyone, I think this is ready to go.

**Pull request merged**